src/logic/div.rs

```diff
@@ -0,0 +1,1290 @@
+/// (quotient) functions, because the division process produces both the quotient and remainder.
+/// However, to stay with Rust's naming scheme we have kept `div` naming. The instruction for
+/// division on many CPUs sets registers to both results of the division process, and compilers will
+/// detect if code uses both results and only use one division instruction. The compiler probably
+/// does not realize this for the `ApInt` division process, and thus the `divrem` and `remdiv` type
+/// instructions exist to explicitly use just one division function for both results.
+/// 
+/// ## Performance
+/// 
+/// All of the division functions in this `impl` quickly check for various edge cases and use an
+/// efficient algorithm for these cases.
+/// Small here means both small ApInt `BitWidth` and/or small **unsigned** numerical significance.
+/// (Signed division works, but two's complement negative numbers may have a large number of
+/// leading ones, leading to potential inefficiency.)
+/// 
+/// - division of zero by any size integer (no allocation)
+/// - division of small (1 `Digit`) integers (no allocation)
+/// - any division that will lead to the quotient being zero or one (no allocation)
+/// - division of any integer by small (1 `Digit`) very small (0.5 `Digit`) integers (no allocation)
+/// - division where the number of leading zeros of both arguments are within one `Digit` (less
+///     allocation than what long division normally requires)
+/// - during long division, the algorithm may encounter a case from above and will use that instead
+/// - division of medium size (<= 512 bits) integers
+/// 
+/// Currently, algorithms faster than 𝒪(n^2) are not implemented, so large integer division may be
+/// very slow compared to other algorithms.
+/// 
+/// Note: Currently there is just one internal division function that is optimized for the
+/// `udivrem` kind of function instead of `uremdiv`. In the future, there is planned a better
+/// implementation for the second one. It was found during designing the first implementation that
+/// this future one should be fundamentally faster than the current one (noticable for small to
+/// medium size `ApInt`s) because of the way `lhs` is subtracted.
+impl ApInt {
+    //Note: the invariant of `ApInt`s where unused bits beyond the bit width must be all zero is
+    //used heavily here, so that no `clear_unused_bits` needs to be used.
+
+    /// This function is intended to be inlined into all of the unsigned quotient and remainder
+    /// functions for optimal assembly.
+    /// `duo` is divided by `div`, and the quotient is assigned to `duo` and remainder assigned
+    /// to `div`
+    /// `false` is returned if division by zero happened. Nothing is modified in the case of
+    /// division by zero.
+    #[inline]
+    pub(crate) fn aarons_algorithm_divrem(duo: &mut [Digit], div: &mut [Digit]) -> bool {
```

Comment:
You might consider a name that relates to the underlying algorithm that is used.